src/compas_rhino/geometry/brep/brep.py

```diff
@@ -70,18 +71,22 @@ def __init__(self, brep=None):
 
     @property
     def data(self):
-        faces = []
-        for face in self.faces:
-            faces.append(face.data)
-        return {"faces": faces}
+        return {
+            "vertices": [v.data for v in self.vertices],
+            "edges": [e.data for e in self.edges],
+            "faces": [f.data for f in self.faces],
+        }
 
     @data.setter
     def data(self, data):
-        faces = []
-        for facedata in data["faces"]:
-            face = RhinoBrepFace.from_data(facedata)
-            faces.append(face)
-        self._create_native_brep(faces)
+        builder = RhinoBrepBuilder()
+        for v_data in data["vertices"]:
+            RhinoBrepVertex.from_data(v_data, builder)
+        for e_data in data["edges"]:
+            RhinoBrepEdge.from_data(e_data, builder)
+        for f_data in data["faces"]:
+            RhinoBrepFace.from_data(f_data, builder)
+        self._brep = builder.result
```

Comment:
would it not make more sense to add stuff to the builder rather than to modify the signature of  all the constructors to send the builder around?

Reply:
The builder serves:
- keeps the reconstruction logic in a single place
- keeps the reconstruction logic separate from the de-serialization
- makes the reconstruction context (i.e. `Brep`) available to each of the components while avoiding global state
- enables the de-serialized elements to hold the native `Rhino.Geometry` type immediately after de-serialization.

`BrepBuilder` is the main context, it essentially wraps around the newly created `Rhino.Geometry.Brep` and holds it until it's properly built.

Since Face, Loop and Trim are nested in each other:

`BrepFaceBuilder` wraps around `Rhino.Geometry.BrepFace` and handed over to the loops so that they can add themselves to the face.
`BrepLoopBuilder` wraps around `Rhino.Geometry.BrepLoop` and handed over to the trims so that they can add themselves to the loop.